openshift/console
# ODC-7232, ODC-7233: Add new auth, serverconfig and usage metrics

# Implements
**Epic [ODC-7171](https://issues.redhat.com/browse/ODC-7171) Improved telemetry (provide new metrics via insight**
* [ODC-7232](https://issues.redhat.com/browse/ODC-7232) Add new metrics and new API to count events to console backend
* [ODC-7233](https://issues.redhat.com/browse/ODC-7233) Track frontend usage and report it to the new metrics API

Aligned documentation:

* [OpenShift UI Telemetry - Prometheus Metrics overview](https://docs.google.com/document/d/1PqbKv_-q2PW8mK3lwGEjpLwO5jdf9TojOchnjUY9YMU/edit) (gdocs)

# Available metrics

## Capability ➔ Is the console installed?

To detect if the console is installed or not you can use this existing (4.12?) metric:

0. `cluster_version_capability` with label `name=Console`

**Query examples:**

```
sum by (name) (cluster_version_capability)

cluster_version_capability{name="Console"}
```

![image](https://user-images.githubusercontent.com/139310/222422541-3e157c9a-9fad-4296-9c2f-24ced3a39b28.png)

## Auth (login/logout) ➔ how many users use the console / what user roles

The login process uses OAuth and counts outgoing redirects from the console and when the user comes "back".

1. `console_auth_login_requests_total` without any label
2. `console_auth_login_successes_total` with label `role` that diff. `kubeadmin`, `cluster-admin` and `developer`
3. `console_auth_login_failures_total` with label `reason`. The reason is currently always `unknown`.
4. `console_auth_logout_requests_total` with label `reason` as well. The reason is currently always `unknown`, too. We could add reasons like `session_timeout` or `user_request` in the future

**Note:** With the default oauth-controller provider we can not "cancel" or fail the login. Other oauth providers might have a "back" button that triggers this.

**Query examples:**

```
sum (rate(console_auth_login_requests_total[1h])) * 3600

sum by (role) (rate(console_auth_login_successes_total[1h])) * 3600

sum by (reason) (rate(console_auth_login_failures_total[1h])) * 3600

sum by (reason) (rate(console_auth_logout_requests_total[1h])) * 3600
```

![image](https://user-images.githubusercontent.com/139310/222438359-db6678b5-a648-4519-90e7-39c434dc95b1.png)

## Usage (page views and impressions) ➔ how many users use the console / what perspective

5. `console_usage_total` metric with labels `event` and `perspective`. Current events are
   * `page_view` is triggered once per "browser load" ~= page_view
   * `page_impression` is used every time the page URL changes, incl. history.push / react-router.push/replace
   * `perspective_changed` when the perspective is changed ;)

**Query examples:**

```
sum by (event, perspective) (console_usage_total)

sum by (event, perspective) (rate(console_usage_total[1h])) * 3600
```

![image](https://user-images.githubusercontent.com/139310/222439386-a4b0fd06-ac83-416b-9a50-f64f0f9fabb3.png)

## Users (console users) ➔ many users use the console / what user roles

6. `console_usage_users` with label `role` that differentiates this roles:
   * `kubeadmin`
   * `cluster-admin`
   * `developer`
   * `unknown` (is only used if any of the permission checks fail)


**Query examples:**

```
max by (role) (console_usage_users)
```

![image](https://user-images.githubusercontent.com/139310/222543449-5d198b87-0476-4aea-be6b-131098e5c6f3.png)

## Plugins ➔ Which plugins are our customers use?

7. `console_plugins_info`  with label (plugin) `name` and `state`. The state is currently always `enabled`. And the counter is always 1.

**Query examples:**

```
avg by (name, state) (console_plugins_info)
```

![image](https://user-images.githubusercontent.com/139310/222452873-72ce5de3-1b26-4008-92ee-76f7fa9466ef.png)

## Customizations (perspectives) ➔ Are customers using this feature?

8. `console_customization_perspectives_info` with label `name` and `state`
   * name is admin, dev, acm, etc.
   * state is one of this: `enabled`, `disabled`, `only-for-cluster-admins`, `only-for-developers`, `custom-permissions`

**Query examples:**

```
avg by (name, state) (console_customization_perspectives_info)
```

![image](https://user-images.githubusercontent.com/139310/222445379-b514343a-5421-4c6a-ae62-d4d4599e54b5.png)

# Tests

## Unit tests

Added a bunch of backend tests:

```bash
go test -count 1 ./pkg/auth ./pkg/serverconfig ./pkg/usage
ok      github.com/openshift/console/pkg/auth   0.007s
ok      github.com/openshift/console/pkg/serverconfig   0.007s
ok      github.com/openshift/console/pkg/usage  0.005s
```

## Manual tests

Trigger logins, and page requests and open the observe > metrics page in the admin perspective. These are cluster metrics that are not shown in the dev perspective.

Small script to trigger usage events "manually":

```bash
server_url=localhost:9000
# or
server_url=$(oc whoami --show-console)

curl "$server_url/metrics/usage" -d '{ "event": "page_view", "perspective": "dev" }'
for i in {1..20}; do
  curl "$server_url/metrics/usage" -d '{ "event": "page_impression", "perspective": "dev" }'
  sleep 3
done
```

**Browser conformance**: 

/cc @debsmita1 @Lucifergene
